src/Components/Components/src/Rendering/ComponentState.cs

```diff
@@ -222,5 +222,27 @@ private void DisposeBuffers()
             ((IDisposable)CurrentRenderTree).Dispose();
             _latestDirectParametersSnapshot?.Dispose();
         }
+
+        internal bool RequiresAsyncDisposal() => Component is IAsyncDisposable;
+
+        internal async Task DisposeInBatchAsync(RenderBatchBuilder batchBuilder)
```

Comment:
It's awkward that this duplicates a bit of delicate logic from `ComponentState.TryDisposeInBatch`. Is it possible to move this logic into `ComponentState` so that it's all in one place? Is it possible to consolidate the sync/async flow a bit more, so maybe there's just one "dispose" method for both, and it returns `Task.CompletedTask` if disposal is synchronous?

Reply:
I did fuse the two, one difference to think about is why do we dispose the component first in the synchronous case. I disposed the component state resources in this case before disposing the component so that they can be released earlier.

Reply:
I extracted the common logic into a separate method